granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -73,28 +95,231 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
     if len(cgroup2_mounts) > 1:
         raise Exception("More than one cgroup2 mount found!")
     path = cgroup2_mounts[0].mount_point
-    if resolve_host_root_links:
-        path = ns.resolve_host_root_links(path)
+    path = ns.resolve_host_root_links(path)
     return path
 
 
-def get_cgroup_mount(controller: str, resolve_host_root_links: bool = True) -> Optional[str]:
+# Cgroup v1 and v2 handle the cgroup assigned processes using cgroup.procs
+CGROUP_PROCS_FILE = "cgroup.procs"
+
+
+class CgroupCore:
+    """
+    This object abstracts the general operations for CGroups for both v1 and v2:
+    - Getting processes that belong to the cgroup or assigning processes to the cgroup.
+    - Reading from or writing to interface files.
+    - Creating CGroups under the hierarchy.
+    cgroup_abs_path: The absolute path for this CGroupCore object
     """
-    Returns the folder that the requested controller is mounted to, or None if no such controller mount was found
+
+    cgroup_abs_path: Path
+    cgroup_mount_path: Path
+
+    def __init__(self, cgroup_abs_path: Path, cgroup_mount_path: Path):
+        self.cgroup_abs_path = cgroup_abs_path
+        self.cgroup_mount_path = cgroup_mount_path
+
+    def _create_subcgroup(self, cgroup_name: str) -> Path:
+        new_cgroup_path = self.cgroup_abs_path / cgroup_name
+        new_cgroup_path.mkdir(exist_ok=True)
+        return new_cgroup_path
+
+    def get_pids_in_cgroup(self) -> set[int]:
+        return {int(proc) for proc in self.read_from_interface_file(CGROUP_PROCS_FILE).split()}
+
+    def assign_process_to_cgroup(self, pid: int = 0) -> None:
+        """
+        :param pid: the pid, if 0 - current process
+        """
+        # This way is consistent between cGroups v1 and v2
+        # The tasks file in cGroups v1 is used for moving specific threads instead of processes
+        self.write_to_interface_file(CGROUP_PROCS_FILE, str(pid))
+
+    def read_from_interface_file(self, interface_name: str) -> str:
+        interface_path = self.cgroup_abs_path / interface_name
+        return interface_path.read_text()
+
+    def write_to_interface_file(self, interface_name: str, data: str) -> None:
+        interface_path = self.cgroup_abs_path / interface_name
+        interface_path.write_text(data)
+
+    @abstractmethod
+    def get_subcgroup(self, controller: ControllerType, cgroup_name: str) -> CgroupCore:
+        """
+        Return a CGroup which has the given name.
+        If current cgroup is of a different name, create a subcgroup with the given name.
+        """
+        pass
+
+    @property
+    def is_v1(self) -> bool:
+        return False
+
+    @property
+    def is_v2(self) -> bool:
+        return False
```

Comment:
If you store the filesystem type in this class, then you don't need to ever override these properties:
```suggestion
    @property
    def is_v1(self) -> bool:
        return self.filesystem_type == 'cgroup'

    @property
    def is_v2(self) -> bool:
        return self.filesystem_type == 'cgroup2'
```

Reply:
I'd still need to override it in the subclaseses, do you prefer it because it's one property instead of 2?